### PR TITLE
*: fix rand method panic

### LINKED ISSUE #1915: PD panic
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

Use TiDB cluster and PD panic in startup.

2. What did you expect to see?

No panic

3. What did you see instead?

```
[2019/11/07 21:32:37.516 +08:00] [INFO] [coordinator.go:211] ["create scheduler with independent configuration"] [scheduler-name=]
[2019/11/07 21:32:37.516 +08:00] [ERROR] [coordinator.go:213] ["can not add scheduler with independent configuration"] [scheduler-name=] [error="scheduler existed"]
[2019/11/07 21:32:37.517 +08:00] [INFO] [coordinator.go:211] ["create scheduler with independent configuration"] [scheduler-name=]
[2019/11/07 21:32:37.517 +08:00] [ERROR] [coordinator.go:213] ["can not add scheduler with independent configuration"] [scheduler-name=] [error="scheduler existed"]
[2019/11/07 21:32:37.518 +08:00] [INFO] [coordinator.go:233] ["create scheduler"] [scheduler-name=balance-region-scheduler]
[2019/11/07 21:32:37.520 +08:00] [INFO] [coordinator.go:233] ["create scheduler"] [scheduler-name=balance-leader-scheduler]
[2019/11/07 21:32:37.521 +08:00] [INFO] [coordinator.go:233] ["create scheduler"] [scheduler-name=balance-hot-region-scheduler]
[2019/11/07 21:32:37.522 +08:00] [INFO] [coordinator.go:233] ["create scheduler"] [scheduler-name=label-scheduler]
[2019/11/07 21:32:37.525 +08:00] [INFO] [coordinator.go:90] ["coordinator starts patrol regions"]
[2019/11/07 21:32:37.525 +08:00] [INFO] [coordinator.go:139] ["coordinator begins to actively drive push operator"]
[2019/11/07 21:32:37.525 +08:00] [FATAL] [log.go:294] [panic] [recover="\"invalid argument to Intn\""] [stack="github.com/pingcap/log.Fatal\n\t/home/jenkins/agent/workspace/build_pd_master/go/pkg/mod/github.com/pingcap/log@v0.0.0-20190715063458-479153f07ebd/global.go:59\ngithub.com/pingcap/pd/pkg/logutil.LogPanic\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/pkg/logutil/log.go:294\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:522\nmath/rand.(*Rand).Intn\n\t/usr/local/go/src/math/rand/rand.go:169\nmath/rand.Intn\n\t/usr/local/go/src/math/rand/rand.go:329\ngithub.com/pingcap/pd/server/core.(*RegionsInfo).GetKeys\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/core/region.go:800\ngithub.com/pingcap/pd/server/core.(*RegionsInfo).RandLeaderRegions\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/core/region.go:770\ngithub.com/pingcap/pd/server/core.(*BasicCluster).RandLeaderRegion\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/core/basic_cluster.go:168\ngithub.com/pingcap/pd/server.(*RaftCluster).RandLeaderRegion\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/cluster.go:632\ngithub.com/pingcap/pd/server/schedulers.(*balanceLeaderScheduler).transferLeaderOut\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/schedulers/balance_leader.go:180\ngithub.com/pingcap/pd/server/schedulers.(*balanceLeaderScheduler).Schedule\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/schedulers/balance_leader.go:148\ngithub.com/pingcap/pd/server.(*scheduleController).Schedule\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/coordinator.go:493\ngithub.com/pingcap/pd/server.(*coordinator).runScheduler\n\t/home/jenkins/agent/workspace/build_pd_master/go/src/github.com/pingcap/pd/server/coordinator.go:446"]
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:56] ["Welcome to Placement Driver (PD)"]
```

[pd-panic.log](https://github.com/pingcap/pd/files/3824060/pd-panic.log)

4. What version of PD are you using (`pd-server -V`)?

```
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:56] ["Welcome to Placement Driver (PD)"]
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:57] [PD] [release-version=v4.0.0-alpha-147-gd6d08d9b]
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:58] [PD] [git-hash=d6d08d9b58fa42a29c3b3f517b9a64a562475588]
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:59] [PD] [git-branch=master]
[2019/11/07 21:32:52.704 +08:00] [INFO] [util.go:60] [PD] [utc-build-time="2019-11-07 09:14:25"]
[2019/11/07 21:32:52.705 +08:00] [INFO] [metricutil.go:80] ["disable Prometheus push client"]
```

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve? 
Fixes #1915. 

### What is changed and how it works?
This PR checks the length of ranges before calling the rand function.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/core/region.go [hunk 1]
```diff
@@ -959,3 +953,12 @@ func (h HexRegionsMeta) String() string {
 	}
 	return strings.TrimSpace(b.String())
 }
+
+// GetKeys gets the start key and end key from random key range.
+func GetKeys(ranges []KeyRange) ([]byte, []byte) {
```
Comment (disksing) [inline server/core/region.go]: Looks not a good name.

File: server/core/region.go [hunk 2]
```diff
@@ -749,58 +749,52 @@ func (r *RegionsInfo) GetStoreLearnerCount(storeID uint64) int {
 
 // RandPendingRegion randomly gets a store's region with a pending peer.
 func (r *RegionsInfo) RandPendingRegion(storeID uint64, ranges []KeyRange) *RegionInfo {
-	startKey, endKey := r.GetKeys(ranges)
+	startKey, endKey := GetKeys(ranges)
```
Comment (disksing) [inline server/core/region.go]: Why not bypass `ranges` to the tree? Randomly pick a range may have problem: if some ranges have no region and the range is randomly selected, the selection process will end with no region selected.

File: server/core/region.go (hunk 1 above)
Comment (shafreeck) [reply server/core/region.go]: Why not name it `RandKeyRange`?

File: server/core/region.go (hunk 2 above)
Comment (rleungx) [reply server/core/region.go]: > Why not bypass ranges to the tree?

Can you explain more?

Comment (disksing) [reply server/core/region.go]: something like
```
sum := 0
result := nil
for _, r := range ranges {
  count := tree.Count(r)
  sum += count
  if rand.Intn(sum) < count {
    result = tree.Rand(r)
  }
}
return result
```

File: server/core/region.go [hunk 3]
```diff
@@ -497,25 +500,51 @@ func (rst *regionSubTree) length() int {
...
 
-func (rst *regionSubTree) RandomRegions(n int, startKey, endKey []byte) []*RegionInfo {
+func (rst *regionSubTree) RandomRegions(n int, ranges []KeyRange) []*RegionInfo {
 	if rst.length() == 0 {
 		return nil
 	}
 
 	regions := make([]*RegionInfo, 0, n)
+
+	if len(ranges) == 0 {
```
Comment (disksing) [inline server/core/region.go]: how about 
```
for i := 0; i <n ; i++ {
  region := rst.RandomRegion(ranges)
  ...
}
```

Comment (lhy1024) [reply server/core/region.go]: what is difference between them?

Comment (disksing) [reply server/core/region.go]: make it shorter I guess

File: server/core/region_tree.go [hunk 4]
```diff
@@ -198,34 +199,49 @@ func (t *regionTree) getAdjacentRegions(region *RegionInfo) (*regionItem, *regio
...
+		}
 
-	if endIndex <= startIndex {
-		log.Error("wrong keys",
-			zap.String("start-key", fmt.Sprintf("%s", HexRegionKey(startKey))),
-			zap.String("end-key", fmt.Sprintf("%s", HexRegionKey(startKey))))
-		return nil
+		if endIndex <= startIndex {
+			log.Error("wrong keys",
+				zap.String("start-key", fmt.Sprintf("%s", HexRegionKey(startKey))),
```
Comment (disksing) [inline server/core/region_tree.go]:
```suggestion
				zap.String("start-key", string(HexRegionKey(startKey))),
```

File: server/core/region_tree.go [hunk 5]
```diff
@@ -198,34 +199,49 @@ func (t *regionTree) getAdjacentRegions(region *RegionInfo) (*regionItem, *regio
...
+				zap.String("end-key", fmt.Sprintf("%s", HexRegionKey(startKey))))
```
Comment (disksing) [inline server/core/region_tree.go]:
```suggestion
				zap.String("end-key", string(HexRegionKey(startKey))))
```

File: server/core/region_tree.go [hunk 6]
```diff
@@ -198,34 +198,49 @@ func (t *regionTree) getAdjacentRegions(region *RegionInfo) (*regionItem, *regio
...
 
-	if endIndex <= startIndex {
-		log.Error("wrong keys",
-			zap.String("start-key", fmt.Sprintf("%s", HexRegionKey(startKey))),
-			zap.String("end-key", fmt.Sprintf("%s", HexRegionKey(startKey))))
-		return nil
+		if endIndex <= startIndex {
+			log.Error("wrong keys",
+				zap.String("start-key", string(HexRegionKey(startKey))),
+				zap.String("end-key", string(HexRegionKey(startKey))))
```
Comment (disksing) [inline server/core/region_tree.go]: startKey -> endKey

**Pull request merged**